pkg/resource/cache_cluster/hooks.go

```diff
@@ -0,0 +1,133 @@
+)
+
+var (
+	condMsgCurrentlyDeleting      = "CacheCluster is currently being deleted"
+	condMsgNoDeleteWhileModifying = "Cannot delete CacheCluster while it is being modified"
+	condMsgCurrentlyUpdating      = "CacheCluster is currently being updated"
+)
+
+var (
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		fmt.Errorf("CacheCluster is in %q state, it cannot be deleted", statusDeleting),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+	requeueWaitWhileModifying = ackrequeue.NeededAfter(
+		fmt.Errorf("CacheCluster is in %q state, it cannot be modified", statusModifying),
+		ackrequeue.DefaultRequeueAfterDuration,
+	)
+)
+
+func hasStatus(r *resource, status string) bool {
+	return r.ko.Status.CacheClusterStatus != nil && *r.ko.Status.CacheClusterStatus == status
+}
+
+func isCreating(r *resource) bool {
+	return hasStatus(r, statusCreating)
+}
+
+func isAvailable(r *resource) bool {
+	return hasStatus(r, statusAvailable)
+}
+
+func isDeleting(r *resource) bool {
+	return hasStatus(r, statusDeleting)
+}
+
+func isModifying(r *resource) bool {
+	return hasStatus(r, statusModifying)
+}
+
+// getTags retrieves the resource's associated tags.
+func (rm *resourceManager) getTags(
+	ctx context.Context,
+	resourceARN string,
+) ([]*svcapitypes.Tag, error) {
+	return util.GetTags(ctx, rm.sdkapi, rm.metrics, resourceARN)
+}
+
+// syncTags keeps the resource's tags in sync.
+func (rm *resourceManager) syncTags(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+) (err error) {
+	return util.SyncTags(ctx, desired.ko.Spec.Tags, latest.ko.Spec.Tags, latest.ko.Status.ACKResourceMetadata, ToACKTags, rm.sdkapi, rm.metrics)
+}
+
+func (rm *resourceManager) updateCacheClusterPayload(input *svcsdk.ModifyCacheClusterInput, desired, latest *resource, delta *ackcompare.Delta) error {
+	desiredSpec := desired.ko.Spec
+	var nodesDelta int64
+	if delta.DifferentAt("Spec.NumCacheNodes") && desired.ko.Spec.NumCacheNodes != nil {
+		numNodes := *latest.ko.Spec.NumCacheNodes
+		if pendingModifications := latest.ko.Status.PendingModifiedValues; pendingModifications != nil &&
+			pendingModifications.NumCacheNodes != nil && *pendingModifications.NumCacheNodes > numNodes {
+			numNodes = *pendingModifications.NumCacheNodes
+		}
+		nodesDelta = numNodes - *desired.ko.Spec.NumCacheNodes
+		if nodesDelta > 0 {
+			for i := numNodes; i > numNodes-nodesDelta; i-- {
+				nodeID := fmt.Sprintf("%04d", i)
+				input.CacheNodeIdsToRemove = append(input.CacheNodeIdsToRemove, &nodeID)
+			}
+		}
+	}
+
+	if idx := slices.IndexFunc(delta.Differences, func(diff *ackcompare.Difference) bool {
+		return diff.Path.Contains("Spec.PreferredAvailabilityZones")
+	}); idx != -1 && desired.ko.Spec.PreferredAvailabilityZones != nil {
+		if nodesDelta >= 0 {
+			return errors.New("spec.preferredAvailabilityZones can only be changed when new nodes are being added via spec.numCacheNodes")
```

Comment:
do you want to make those terminal errors?

Reply:
It's a terminal error. Any error returned from this function is treated as a terminal error.